resolver.go

```diff
@@ -373,11 +391,13 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 			}()
 			err = co.WriteMsg(query)
 			if err != nil {
+				r.concurrentQueryDec()
```

Comment:
What I am trying to say is that once the code reaches line 386, it will decrement the counter anyway.
It does not matter that it does it after an unsuccessful `co.WriteMsg(query)` or after the call to `co.ReadMsg()`.
The logic seems to be that once `r.concurrentQueryInc() == true`, `r.concurrentQueryDec()` will be executed no matter what.

So I was suggesting to remove the two calls to `r.concurrentQueryDec()` at line 394 and 400 and add one at line 386.